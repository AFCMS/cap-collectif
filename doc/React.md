### How to write localizable components

Just import the appropriate [component](https://github.com/yahoo/react-intl/wiki#the-react-intl-module) from `react-intl`

- For localizable text use
[`<FormattedMessage>`](https://github.com/yahoo/react-intl/wiki/Components#formattedmessage).

- For date and time:
[`<FormattedDate>`](https://github.com/yahoo/react-intl/wiki/Components#formatteddate)
[`<FormattedTime>`](https://github.com/yahoo/react-intl/wiki/Components#formattedtime)
[`<FormattedRelative>`](https://github.com/yahoo/react-intl/wiki/Components#formattedrelative)

- For numbers and currencies:
[`<FormattedNumber>`](https://github.com/yahoo/react-intl/wiki/Components#formattednumber)
[`<FormattedPlural>`](https://github.com/yahoo/react-intl/wiki/Components#formattedplural)

- If possible, do not use `<FormattedHTMLMessage>`, see how to use *Rich Text Formatting* with
[`<FormattedMessage>`](https://github.com/yahoo/react-intl/wiki/Components#formattedmessage)

- When you need an imperative formatting API, use the [`injectIntl`](https://github.com/yahoo/react-intl/wiki/API#injectintl) High-Order Component.
