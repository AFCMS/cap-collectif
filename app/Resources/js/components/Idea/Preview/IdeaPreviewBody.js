import React from 'react';
import { FormattedMessage } from 'react-intl';

const IdeaPreviewBody = React.createClass({
  propTypes: {
    idea: React.PropTypes.object.isRequired,
  },

  render() {
    const { idea } = this.props;
    return (
      <div className="card__body">
        <div className="card__body__infos">
          <h3 className="card__title">
            <a href={idea._links.show}>{idea.title}</a>
          </h3>
          {idea.trashed ? (
            <span className="label label-default">
            {<FormattedMessage id="idea.preview.trashed" />}
          </span>
          ) : null}
        </div>
        <span className="excerpt small">
          <FormattedMessage
            id="idea.preview.counters.votes"
            values={{
              num: idea.votesCount,
            }}
          />
          {idea.commentable ? (
            <span>
              {' â€¢ '}
              <FormattedMessage
                id="idea.preview.counters.comments"
                values={{
                  num: idea.commentsCount,
                }}
              />
            </span>
          ) : null}
        </span>
      </div>
    );
  },
});

export default IdeaPreviewBody;
