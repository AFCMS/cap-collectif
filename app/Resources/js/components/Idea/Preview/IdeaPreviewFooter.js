import React from 'react';
import { FormattedMessage } from 'react-intl';

const IdeaPreviewFooter = React.createClass({
  propTypes: {
    idea: React.PropTypes.object.isRequired,
  },

  render() {
    const { idea } = this.props;
    return (
      <div className="idea__preview__footer">
        <span className="excerpt small">
          <FormattedMessage
            id="idea.preview.counters.votes"
            values={{
              num: idea.votesCount,
            }}
          />
          {idea.commentable ? (
            <span>
              {' â€¢ '}
              <FormattedMessage
                id="idea.preview.counters.comments"
                values={{
                  num: idea.commentsCount,
                }}
              />
            </span>
          ) : null}
        </span>
      </div>
    );
  },
});

export default IdeaPreviewFooter;
