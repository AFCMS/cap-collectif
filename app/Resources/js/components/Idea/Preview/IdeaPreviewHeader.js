import React from 'react';
import moment from 'moment';
import { FormattedDate } from 'react-intl';
import UserAvatar from '../../User/UserAvatar';
import UserLink from '../../User/UserLink';
import { CardUser } from '../../Ui/Card/CardUser';

const IdeaPreviewHeader = React.createClass({
  propTypes: {
    idea: React.PropTypes.object.isRequired,
  },

  render() {
    const { idea } = this.props;
    return (
      <CardUser>
        <div className="card__user__avatar">
          <UserAvatar user={idea.author} className="pull-left idea__avatar" />
        </div>
        <div className="ellipsis">
          <UserLink user={idea.author} className="small" />
          <p className="excerpt small">
            <FormattedDate
              value={moment(idea.createdAt)}
              day="numeric"
              month="long"
              year="numeric"
              hour="numeric"
              minute="numeric"
            />
          </p>
        </div>
        <hr />
      </CardUser>
    );
  },
});

export default IdeaPreviewHeader;
