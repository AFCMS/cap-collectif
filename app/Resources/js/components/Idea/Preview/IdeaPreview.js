import React from 'react';
import classNames from 'classnames';
import IdeaPreviewHeader from './IdeaPreviewHeader';
import IdeaPreviewBody from './IdeaPreviewBody';
import { CardContainer } from '../../Ui/Card/CardContainer';

const IdeaPreview = React.createClass({
  propTypes: {
    idea: React.PropTypes.object.isRequired,
  },

  render() {
    const { idea } = this.props;
    const classes = classNames({
      'bg-vip': idea.author.vip,
    });
    return (
      <CardContainer className={classes}>
        <IdeaPreviewHeader idea={idea} />
        <IdeaPreviewBody idea={idea} />
      </CardContainer>
    );
  },
});

export default IdeaPreview;
