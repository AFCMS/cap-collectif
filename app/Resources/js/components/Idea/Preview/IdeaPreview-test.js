/* eslint-env jest */
/* eslint no-unused-expressions:0 */
import React from 'react';
import { expect } from 'chai';
import { shallow } from 'enzyme';
import IntlData from '../../../translations/FR';
import IdeaPreview from './IdeaPreview';
import IdeaPreviewHeader from './IdeaPreviewHeader';
import IdeaPreviewBody from './IdeaPreviewBody';
import IdeaPreviewFooter from './IdeaPreviewFooter';

const idea = {
  id: 1,
  author: {},
};

const ideaVip = {
  id: 1,
  author: {
    vip: true,
  },
};

describe('<IdeaPreview />', () => {
  it('it should render idea preview', () => {
    const wrapper = shallow(<IdeaPreview idea={idea} {...IntlData} />);
    expect(wrapper.find('.idea__preview')).to.have.length(1);
    expect(wrapper.find('.idea__preview.bg-vip')).to.have.length(0);
    expect(wrapper.find(IdeaPreviewHeader)).to.have.length(1);
    expect(wrapper.find(IdeaPreviewBody)).to.have.length(1);
    expect(wrapper.find(IdeaPreviewFooter)).to.have.length(1);
  });

  it('should render vip background when idea author is vip', () => {
    const wrapper = shallow(<IdeaPreview idea={ideaVip} {...IntlData} />);
    expect(wrapper.find('.idea__preview.bg-vip')).to.have.length(1);
  });
});
