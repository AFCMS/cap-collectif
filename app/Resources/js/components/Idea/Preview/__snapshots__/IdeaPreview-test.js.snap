// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IdeaPreview /> it should render idea preview 1`] = `
<CardContainer
  className=""
>
  <IdeaPreviewHeader
    idea={
      Object {
        "author": Object {},
        "id": 1,
      }
    }
  />
  <IdeaPreviewBody
    idea={
      Object {
        "author": Object {},
        "id": 1,
      }
    }
  />
</CardContainer>
`;

exports[`<IdeaPreview /> should render vip background when idea author is vip 1`] = `
<CardContainer
  className="bg-vip"
>
  <IdeaPreviewHeader
    idea={
      Object {
        "author": Object {
          "vip": true,
        },
        "id": 1,
      }
    }
  />
  <IdeaPreviewBody
    idea={
      Object {
        "author": Object {
          "vip": true,
        },
        "id": 1,
      }
    }
  />
</CardContainer>
`;
